File: NodeDesignPatterns/iterator.js

doc =`


+-----------+
| Aggregate |
+-----------+
| data      |
+-----------+
| iterator()|
+-----+-----+
      |
      | creates
      v
+-----------+
| Iterator  |
+-----------+
| position  |
+-----------+
| next()    |
+-----------+


`

console.log(doc)


class PlayListIterator{
    constructor(songs){
        this.songs=songs
        this.i=0
    }
    next(){

        if (this.i<this.songs.length){
            this.i+=1
            return {value:this.songs[this.i-1], done:false}
        }
        else{
            return {done:true}
        }

    }
}


class PlayList{

    constructor(){
        this.songs=[]
    }

    add_song(song){
        this.songs.push(song)
    }
    [Symbol.iterator](){
        return new PlayListIterator(this.songs)
    }
}




let music_list=new PlayList()

music_list.add_song("Baby doll")
music_list.add_song("Mona Mona")
music_list.add_song("Afreen Afreen")
music_list.add_song("Robo")


for (let i of music_list){
    console.log(i)
}

File: PythonDesignPatterns/iterator.py

"""
+-----------+
| Aggregate |
+-----------+
| data      |
+-----------+
| iterator()|
+-----+-----+
      |
      | creates
      v
+-----------+
| Iterator  |
+-----------+
| position  |
+-----------+
| next()    |
+-----------+


"""

print(__doc__)


class PlaylistIterator:
    def __init__(self,songs):
        self.songs=songs
        self.i=0
    def __next__(self):

        if self.i < len(self.songs):
            self.i+=1
            return self.songs[self.i-1]
        raise StopIteration


        



class Playlist:

    def __init__(self):
        self.songs=[]
    

    def add_song(self,song):
        self.songs.append(song)
    
    def __iter__(self):
        return PlaylistIterator(self.songs)
    



music_list=Playlist()

music_list.add_song("Baby doll")
music_list.add_song("Beinteha")
music_list.add_song("Mona Mona")
music_list.add_song("Afreen Afreen")
music_list.add_song("Robo")


for i  in music_list:
    print(i)
